azure/src/main/scala/quasar/destination/avalanche/azure/AvalancheAzureDestination.scala

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright 2020 Precog Data
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package quasar.destination.avalanche.azure
+
+import quasar.destination.avalanche.AvalancheQueries._
+
+
+import cats.ApplicativeError
+import cats.data.NonEmptyList
+import cats.effect.concurrent.Ref
+import cats.effect.{ConcurrentEffect, ContextShift, Sync, Timer}
+import cats.implicits._
+import com.azure.storage.blob.BlobContainerAsyncClient
+import doobie._
+import doobie.free.connection.createStatement
+import doobie.implicits._
+import fs2.{compression, Stream}
+import java.util.UUID
+import java.net.URI
+import org.slf4s.Logging
+import pathy.Path.FileName
+import quasar.api.destination.DestinationType
+import quasar.api.resource._
+import quasar.api.ColumnType
+import quasar.blobstore.azure.{AzureDeleteService, AzurePutService, Expires, TenantId}
+import quasar.blobstore.paths.{BlobPath, PathElem}
+import quasar.connector.destination.{LegacyDestination, ResultSink}
+import quasar.connector.{MonadResourceErr, ResourceError}
+import quasar.destination.avalanche.WriteMode._
+import scala.{
+  Byte,
+  Int,
+  List,
+  RuntimeException,
+  Seq,
+  Some,
+  StringContext,
+  Throwable,
+  Unit
+ }
+import scala.Predef.String
+import scala.util.matching.Regex
+
+final class AvalancheAzureDestination[F[_]: ConcurrentEffect: ContextShift: MonadResourceErr: Timer](
+    xa: Transactor[F],
+    refContainerClient: Ref[F, Expires[BlobContainerAsyncClient]],
+    refreshToken: F[Unit],
+    config: AvalancheAzureConfig) extends LegacyDestination[F] with Logging {
+
+  def destinationType: DestinationType =
+    AvalancheAzureDestinationModule.destinationType
+
+  def sinks: NonEmptyList[ResultSink[F, ColumnType.Scalar]] = NonEmptyList.of(csvSink)
+
+  private val csvSink: ResultSink[F, ColumnType.Scalar] = 
+    ResultSink.create[F, ColumnType.Scalar](AvalancheRenderConfig) {
+      case (path, columns, bytes) =>
+        for {
+          tableName <- Stream.eval(ensureValidTableName(path))
+
+          cols <- ApplicativeError[Stream[F, ?], Throwable].fromEither(
+            ensureValidColumns(columns).leftMap(new RuntimeException(_)))
+
+          compressed = bytes.through(compression.gzip(bufferSize = 1024 * 32))
+
+          freshName <- Stream.eval(upload(compressed))
+
+          _ <- Stream.eval(copy(tableName, freshName, cols)).onFinalize(deleteBlob(freshName))
```

Comment:
NB this should be done with `bracket` rather than `eval` and `onFinalize`, since we will leak the file if we are canceled during the upload.